internal/api/core/routing.go

```diff
@@ -152,7 +152,22 @@ func sendAPIResponse(apiRequest ValidAPIRequest, response http.ResponseWriter,
 	}
 
 	endpoint, sender, userEmail, courseID, assignmentID, locator := getRequestInfo(apiRequest, apiErr)
-	stats.AsyncStoreAPIRequestMetric(startTime, apiResponse.EndTimestamp, sender, endpoint, userEmail, courseID, assignmentID, locator)
+	metric := stats.Metric{
+		Timestamp: startTime,
+		Type:      stats.API_REQUEST_STATS_TYPE,
+		Value:     float64((apiResponse.EndTimestamp - startTime).ToMSecs()),
+		Attributes: map[stats.MetricAttribute]any{
+			stats.SENDER_KEY:   sender,
+			stats.ENDPOINT_KEY: endpoint,
+		},
+	}
+
+	stats.InsertIntoMapIfPresent(metric.Attributes, stats.USER_EMAIL_KEY, userEmail)
+	stats.InsertIntoMapIfPresent(metric.Attributes, stats.COURSE_ID_KEY, courseID)
+	stats.InsertIntoMapIfPresent(metric.Attributes, stats.ASSIGNMENT_ID_KEY, assignmentID)
+	stats.InsertIntoMapIfPresent(metric.Attributes, stats.LOCATOR_KEY, locator)
```

Comment:
Consider having methods to set each of these values so you don't have to supply the key each time.
E.g., `metric.SetCourse(courseID)`.

Note that this is a soft suggestion ("consider").
You have a better mental image of the code, so think about it.